content/blog/elm-review-v2/index.md

```diff
@@ -0,0 +1,269 @@
+
+For instance you can write rules that
+
+- report when you [pass values outside of 0-255 to `Css.rgb255`](https://package.elm-lang.org/packages/folq/review-rgb-ranges/latest/)
+- give an error if any `Regex.fromString` calls have invalid regexes (for literal strings)
+
+Unfortunately, as soon as a problem spawns multiple modules, you wouldn't be able to enforce those guarantees. But that was the case until we got...
+
+## Project rules!
+
+In `elm-review` version 1, `Elm Analyse`, `ESLint` and a lot of similar tools for other languages, the analysis is scoped to a single file.
+
+This means that when a rule has finished looking at module A and starts looking at module B, it forgets everything about module A. That means we can't answer a lot of questions that we would like the answer to, such as:
+
+- Which module does function `xyz` come from? If a module is imported using `import A exposing (..)`, then we potentially lose the ability to tell. We have the same problem with types when encountering `import A exposing (B(..))`.
+- What is the type of an imported function?
+- Is this element ever used in the project?
+
+In addition to the **module rules** that we just described, `elm-review` now has **project rules**. These rules go through all the modules and can use information collected from a different module to infer or report things, solving the problems mentioned above.
+
+In short, this feature makes rules much more accurate, and allow for almost any of them to be _complete_. No more "best effort" due to the limitations of looking at a single module.
+
+This opens up a wide range of possibilities. A few example use-cases:
+
+- [Report custom type constructors that aren't used anywhere in the project](https://package.elm-lang.org/packages/jfmengels/review-unused/latest/NoUnused-CustomTypeConstructors)
+- [Report functions/types that are exposed but are never used in the project](https://package.elm-lang.org/packages/jfmengels/review-unused/latest/NoUnused-Exports)
+- [Report unused modules](https://package.elm-lang.org/packages/jfmengels/review-unused/latest/NoUnused-Modules)
+- Report unused fields in records
+- Report when `Html.lazy` is used incorrectly
+- Report when a variable that should contain all the possible variants of a custom type is missing a variant
+- Report when a module uses another module's `update` function but not the `subscription` function
+- Report unused CSS classes from your CSS files (would require you to generate an Elm file from the CSS files in the configuration folder)
+
+## Reporting errors for elm.json
+
+In version 1, reading the contents of `elm.json` was possible, but it wasn't possible to report errors for it.
+
+## Reading the dependencies' docs.json file
+
+The `docs.json` file for the dependencies contain all the public information about your direct and indirect dependencies. Reading these can tell you which dependency a module comes from, along what types and functions they contain.
```

Comment:
"Getting this information from the `docs.json` file is much faster and simpler than parsing all the Elm code in the entire dependency tree for information" or some such might be nice 🙂

Reply:
To do it the parsing way, you'd basically need to clone the repos for every dependency :sweat_smile:

Reply:
Isn't the Elm code part of the package that gets downloaded? Or is that just something I dreamt up? 😛

Reply:
Yeah, it is downloaded at some point, you're right.
I have no clue about how that is done and where that endpoint lies though. Since it's undocumented (unlike the `docs.json`), I prefer not to use that. But yes, you're right :+1:

Reply:
FYI, if the `docs.json` are not found locally, `elm-review` downloads it from the packages website. I have no clue how you could download the files though (I don't know if there is a separate endpoint on the packages' API, or whether it downloads things manually from GitHub)